File: cpp/451. Sort Characters By Frequency.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

// class Solution {
// // Custom sort on frequency, Time: O(n+26log26), Space: O(26), Suppose size of characters is constant.
// public:
//     string frequencySort(string s) {
//         unordered_map<char, int> counter;
//         for(char ch : s) counter[ch]++;
//         vector<pair<char, int>> data(counter.begin(), counter.end()); // [(ch, freq)]
//         auto comp = [](pair<char, int> &lhs, pair<char, int> &rhs) {
//             return lhs.second > rhs.second;
//         };
//         sort(data.begin(), data.end(), comp);

//         string ans = "";
//         for(auto [ch, freq] : data) {
//             for(int i = 0; i < freq; i++) ans.push_back(ch);
//         }
//         return ans;
//     }
// };

class Solution {
// Bucket Sort because freq will never greater than n, Time: O(n), Space: O(n)
public:
    string frequencySort(string s) {
        int n = s.size();
        unordered_map<char, int> counter;
        for(char ch : s) counter[ch]++;
        vector<vector<char>> bucket(n+1, vector<char>());
        for(auto [ch, freq] : counter) bucket[freq].push_back(ch);
        string ans = "";
        for(int freq = n; freq >= 0; freq--) {
            for(char ch : bucket[freq]) for(int j = 0; j < freq; j++) ans.push_back(ch);
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; string s, ans;

    s = "tree";
    ans = solution.frequencySort(s);
    assert(ans=="eert");

    s = "cccaaa";
    ans = solution.frequencySort(s);
    assert(ans=="cccaaa" || ans=="aaaccc");

    s = "Aabb";
    ans = solution.frequencySort(s);
    assert(ans=="bbAa" || ans=="bbaA");
}

File: python/451-Sort Characters By Frequency.py
from collections import Counter

class Solution:
    def frequencySort(self, s: str) -> str:
        "HashMap + Custom sort, Time: O(nlogn), Space: O(n)"
        counter = Counter(s) # O(n)
        result = sorted(s, key=lambda x: (-counter[x], x)) # O(nlogn)
        return "".join(result) # O(n)
    def frequencySort(self, s: str) -> str:
        "HashMap + Heap, Time: O(n), Space: O(n)"
        import heapq
        from collections import defaultdict
        counter = defaultdict(lambda: 0)
        for ch in s: # O(n)
            counter[ch] += 1
        max_heap = []
        for ch, count in counter.items(): # O(26)
            heapq.heappush(max_heap, (-count, ch)) # O(log26)
        result = []
        while max_heap: # O(26)
            minus_count, ch = heapq.heappop(max_heap) # O(log26)
            count = -minus_count
            result.append(ch*count)
        return "".join(result) # O(n)
    def frequencySort(self, s: str) -> str:
        "Counter + most_common, Time: O(n), Space: O(n)"
        result = [ch*count for ch, count in Counter(s).most_common()] # O(n + 26*log26)
        return "".join(result)
s = "tree"
result = Solution().frequencySort(s)
print(result) # "eert"

s = "cccaaa"
result = Solution().frequencySort(s)
print(result) # "cccaaa"

s = "Aabb"
result = Solution().frequencySort(s)
print(result) # "bbAa"

s = "loveleetcode"
result = Solution().frequencySort(s)
print(result) # "eeeeoollvtdc"